src/H5Dmpio.c

```diff
@@ -5578,274 +5420,146 @@ H5D__mpio_get_chunk_insert_info_types(MPI_Datatype *contig_type, hbool_t *contig
-                     * set the base chunk offset and base chunk
-                     * data buffer if we haven't already
-                     */
-                    if (!H5_addr_defined(base_offset)) {
-#if H5_CHECK_MPI_VERSION(3, 0)
-                        if (MPI_SUCCESS != (mpi_code = MPI_Get_address(chunk_list[i].buf, &base_buf)))
-                            HMPI_GOTO_ERROR(FAIL, "MPI_Get_address failed", mpi_code)
-#else
-                        base_buf = (MPI_Aint)chunk_list[i].buf;
-#endif
+            assert(vec_idx < iovec_count);
 
-                        base_offset = chunk_list[i].chunk_current.offset;
-                    }
-                }
-
-                /* Set convenience pointer for current chunk block */
-                chunk_block =
-                    (op_type == H5D_IO_OP_READ) ? &chunk_list[i].chunk_current : &chunk_list[i].chunk_new;
+            if (op_type == H5D_IO_OP_READ && !chunk_list->chunk_infos[i].need_read)
+                continue;
 
-                /*
-                 * Set the current chunk entry's offset in the file, relative to
-                 * the first chunk entry
-                 */
-                assert(H5_addr_defined(chunk_block->offset));
-                file_offset_array[chunk_count] = (MPI_Aint)(chunk_block->offset - base_offset);
+            /* Set convenience pointer for current chunk block */
+            chunk_block = (op_type == H5D_IO_OP_READ) ? &chunk_list->chunk_infos[i].chunk_current
+                                                      : &chunk_list->chunk_infos[i].chunk_new;
 
-                /*
-                 * Ensure the chunk list is sorted in ascending ordering of
-                 * offset in the file
-                 */
-                if (chunk_count)
-                    assert(file_offset_array[chunk_count] > file_offset_array[chunk_count - 1]);
+            assert(H5_addr_defined(chunk_block->offset));
+            io_addrs[vec_idx] = chunk_block->offset;
 
-                /* Set the current chunk entry's size for the I/O operation */
-                H5_CHECK_OVERFLOW(chunk_block->length, hsize_t, int);
-                length_array[chunk_count] = (int)chunk_block->length;
-
-                /*
-                 * Set the displacement of the chunk entry's chunk data buffer,
-                 * relative to the first entry's data buffer
-                 */
-#if H5_CHECK_MPI_VERSION(3, 1)
-                if (MPI_SUCCESS != (mpi_code = MPI_Get_address(chunk_list[i].buf, &chunk_buf)))
-                    HMPI_GOTO_ERROR(FAIL, "MPI_Get_address failed", mpi_code)
+            /*
+             * Ensure the chunk list is sorted in ascending ordering of
+             * offset in the file. Note that we only compare that the
+             * current address is greater than the previous address and
+             * not equal to it; file addresses should only appear in the
+             * chunk list once.
+             */
+            if (vec_idx > 0)
```

Comment:
I assume the compiler will generally optimize this out, but my preference would be to surround this in an #ifndef NDEBUG directive

Reply:
Fixed